File: backend/app/orchestration/skill_sorting.py
"""
技能排序模块。

根据技能来源（内置/已安装插件/本地/独立）与插件名、技能名，
为技能列表提供统一稳定的展示排序规则。
"""

from app.orchestration.skill_registry import SkillMetadata


def get_plugin_type(skill: SkillMetadata) -> str:
    """获取技能的插件类型

    判断规则：
        - 无 plugin_name：视为独立技能 'independent'
        - 有 plugin_name 但无 install_path：视为本地 'local'
        - install_path 包含 '/.reflexion/'：视为已安装插件 'installed'
        - install_path 以 '/skills' 结尾：视为内置 'builtin'
        - 其余情况：'local'

    Args:
        skill: 技能元数据

    Returns:
        str: 插件类型标识，取值为 'builtin'/'installed'/'local'/'independent'
    """
    if not skill.plugin_name:
        return 'independent'
    if not skill.install_path:
        return 'local'

    # 标准化路径
    normalized_path = skill.install_path.replace('\\', '/')
    if '/.reflexion/' in normalized_path:
        return 'installed'
    if normalized_path.endswith('/skills') or normalized_path.endswith('/skills/'):
        return 'builtin'
    return 'local'


def sort_skills(skills: list[SkillMetadata]) -> list[SkillMetadata]:
    """
    排序技能列表
    排序规则：插件类型（builtin < installed < local < independent）→ 插件名 → 技能名

    Args:
        skills: 待排序的技能元数据列表

    Returns:
        list[SkillMetadata]: 按规则排序后的新列表
    """
    type_order = {
        'builtin': 0,
        'installed': 1,
        'local': 2,
        'independent': 3,
    }

    def sort_key(skill: SkillMetadata) -> tuple:
        """计算单个技能的排序键：(类型优先级, 插件名, 技能名)"""
        plugin_type = get_plugin_type(skill)
        type_priority = type_order.get(plugin_type, 999)
        plugin_name = skill.plugin_name or ''
        skill_name = skill.name
        return (type_priority, plugin_name, skill_name)

    return sorted(skills, key=sort_key)

File: frontend/src/utils/skillSorting.ts
import type { Skill } from '@/types/skill'

type PluginTypeKey = 'builtin' | 'installed' | 'local' | 'independent'

const PLUGIN_TYPE_ORDER: Record<PluginTypeKey, number> = {
  builtin: 0,
  installed: 1,
  local: 2,
  independent: 3,
} as const

export type PluginInfo = {
  name: string
  displayName: string
  type: PluginTypeKey
  skillCount: number
}

/**
 * 获取技能的插件类型
 */
export function getPluginType(skill: Skill): PluginTypeKey {
  if (!skill.plugin_name) return 'independent'
  if (!skill.install_path) return 'local'

  // 更精确的路径匹配
  const normalizedPath = skill.install_path.replace(/\\/g, '/')
  if (normalizedPath.includes('/.reflexion/')) return 'installed'
  if (normalizedPath.match(/\/skills\/?$/)) return 'builtin'
  return 'local'
}

/**
 * 获取技能的插件显示名称
 */
export function getPluginDisplayName(skill: Skill): string {
  if (!skill.plugin_name) return '独立技能'
  return skill.plugin_name
}

/**
 * 获取所有插件信息并按优先级排序
 */
export function getPluginList(skills: Skill[]): PluginInfo[] {
  const pluginMap = new Map<string, PluginInfo>()

  // 统计每个插件的技能数量
  skills.forEach((skill) => {
    const type = getPluginType(skill)
    const name = skill.plugin_name || 'independent'
    const displayName = getPluginDisplayName(skill)

    if (!pluginMap.has(name)) {
      pluginMap.set(name, { name, displayName, type, skillCount: 0 })
    }
    const plugin = pluginMap.get(name)
    if (plugin) {
      plugin.skillCount++
    }
  })

  const plugins = Array.from(pluginMap.values())

  // 排序：内置 → 全局安装 → 本地 → 独立，同类型内按技能数量降序
  plugins.sort((a, b) => {
    const typeCompare = PLUGIN_TYPE_ORDER[a.type] - PLUGIN_TYPE_ORDER[b.type]
    if (typeCompare !== 0) return typeCompare
    return b.skillCount - a.skillCount
  })

  return plugins
}

/**
 * 获取优先显示的插件（前3-4个常用插件）
 */
export function getTopPlugins(plugins: PluginInfo[]): PluginInfo[] {
  // 独立技能如果存在，始终显示
  const independent = plugins.find((p) => p.type === 'independent')
  const others = plugins.filter((p) => p.type !== 'independent')

  // 取前3个非独立插件
  const topOthers = others.slice(0, 3)

  return independent ? [...topOthers, independent] : topOthers
}

/**
 * 对技能列表排序
 * 按插件类型 → 插件名 → 技能名
 */
export function sortSkills(skills: Skill[]): Skill[] {
  return [...skills].sort((a, b) => {
    const typeA = getPluginType(a)
    const typeB = getPluginType(b)
    const typeCompare = PLUGIN_TYPE_ORDER[typeA] - PLUGIN_TYPE_ORDER[typeB]
    if (typeCompare !== 0) return typeCompare

    const pluginA = a.plugin_name || ''
    const pluginB = b.plugin_name || ''
    const pluginCompare = pluginA.localeCompare(pluginB)
    if (pluginCompare !== 0) return pluginCompare

    return a.name.localeCompare(b.name)
  })
}
